### PR TITLE
Sema: Fix crash when attempting to diagnose a contextual conversion o…

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@rudkx Do you mind taking a quick look (and pre-emptively sign off on this for 3.0-branch too)?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: validation-test/compiler_crashers_fixed/28385-swift-constraints-constraintgraph-addconstraint.swift [hunk 1]
```diff
@@ -5,6 +5,5 @@
 // See http://swift.org/LICENSE.txt for license information
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: not --crash %target-swift-frontend %s -parse
-// REQUIRES: asserts
+// RUN: not %target-swift-frontend %s -parse
 assert((n:String{
```
Comment (NachoSoto) [inline validation-test/compiler_crashers_fixed/28385-swift-constraints-constraintgraph-addconstraint.swift]: Haha, I fail to see how this is related, but it's funny.

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -4127,6 +4127,12 @@ bool FailureDiagnosis::diagnoseContextualConversionError() {
                                                 CS))
     return true;
 
+  // Don't attempt fixits if we have an unsolved type variable, since
+  // the recovery path's recursion into the type checker via typeCheckCast()
+  // will confuse matters.
+  if (exprType->hasTypeVariable())
```
Comment (NachoSoto) [inline lib/Sema/CSDiag.cpp]: Fancy!!

File: test/Constraints/diagnostics.swift [hunk 3]
```diff
@@ -780,3 +780,18 @@ struct rdar27891805 {
 try rdar27891805(contentsOfURL: nil, usedEncoding: nil)
 // expected-error@-1 {{argument labels '(contentsOfURL:, usedEncoding:)' do not match any available overloads}}
 // expected-note@-2 {{overloads for 'rdar27891805' exist with these partially matching parameter lists: (contentsOf: String, encoding: String), (contentsOf: String, usedEncoding: inout String)}}
+
+// Make sure RawRepresentable fix-its don't crash in the presence of type variables
```
Comment (NachoSoto) [inline test/Constraints/diagnostics.swift]: ❤️

**Review by NachoSoto — APPROVED** [review]
YAY

File: validation-test/compiler_crashers_fixed/28385-swift-constraints-constraintgraph-addconstraint.swift (hunk 1 above)
Comment (slavapestov) [inline validation-test/compiler_crashers_fixed/28385-swift-constraints-constraintgraph-addconstraint.swift]: It produces 'expression type '(n: _)' is ambiguous without more context' now, which sucks, but better than a crash at least.

**Comment by rudkx** [comment]
@slavapestov LGTM. I wonder if these are dups:

https://bugs.swift.org/browse/SR-2547 (rdar://problem/28141573)
rdar://problem/28262370
rdar://problem/28262540

Also, I cannot imagine a case where it would make sense to call into typeCheckCheckedCast (or a variety of other functions) with TypeVariableTypes (or ErrorType). If you have time it would be great to add an assert to that effect!

**Pull request merged**